### PR TITLE
Prototype authentication

### PR DESCRIPTION
**Pull request by shiroyuki**
* [x] Changed from `AZUL_FUSILLADE_HOSTNAME` (just a hostname) to `AZUL_FUSILLADE_ENDPOINT` (base URL)
* [x] Fixed the merging issue
* [x] Fixed the side effect issue.

### CODE REVIEW

**Comment by shiroyuki** [comment]
@hannes-ucsc, in 0ffec77, I'm not sure how you would like me avoid defining side effects at import time. So, I wrap this in a function so that the test can patch the method (`shared_session`).

File: src/azul/security/auth.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+from logging import getLogger
+from typing import Dict, Any
+from urllib.parse import urlencode
+
+from azul import config
+from azul.security.verifier import (verify,
+                                    InvalidTokenError,
+                                    NonDecodableTokenError)
+
+logger = getLogger(__name__)
```
Comment (hannes-ucsc) [inline src/azul/security/auth.py]:
```suggestion
logger = logging.getLogger(__name__)
```

Like everywhere else.

File: src/azul/security/auth.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+fusillade_callback_url = config.service_endpoint() + '/auth/callback'
```
Comment (hannes-ucsc) [inline src/azul/security/auth.py]: This seems to be factored into a constant for patching during testing. I don't like it because it essentially caches  the return value of `config.service_endpoint` when the module is imported. Changes to the underlying environment variables from which `config.service_endpoint` is derived will not be reflected. This is why everything in Config is a property and Config is stateless. This type of caching breaks that.

File: src/azul/security/verifier.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+from azul import config
+
+logger = logging.getLogger(__name__)
+
+allowed_algorithms = ('RS256',)
+gserviceaccount_domain = "iam.gserviceaccount.com"
+
+# Shared HTTP session
+def shared_session():
```
Comment (hannes-ucsc) [inline src/azul/security/verifier.py]: The name of the method and the (redundant) comment are misleading. There is no sharing going on here. Every invocation of a function in this module will create a new session.

File: src/azul/security/verifier.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+import jwt
+from jwt.exceptions import DecodeError, PyJWTError
+import requests
+
+from azul import config
+
+logger = logging.getLogger(__name__)
+
+allowed_algorithms = ('RS256',)
+gserviceaccount_domain = "iam.gserviceaccount.com"
```
Comment (hannes-ucsc) [inline src/azul/security/verifier.py]: Add to Config or inline.

File: src/azul/security/verifier.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+from cryptography.hazmat.backends import default_backend
+import jwt
+from jwt.exceptions import DecodeError, PyJWTError
+import requests
+
+from azul import config
+
+logger = logging.getLogger(__name__)
+
+allowed_algorithms = ('RS256',)
```
Comment (hannes-ucsc) [inline src/azul/security/verifier.py]: Add to Config or inline.

File: test/app_test_case.py [hunk 6]
```diff
@@ -16,6 +25,37 @@
 from azul_test_case import AzulTestCase
 
 log = logging.getLogger(__name__)
+key_dir = os.path.dirname(__file__)
```
Comment (hannes-ucsc) [inline test/app_test_case.py]: Use tempdir instead.

File: test/app_test_case.py [hunk 7]
```diff
@@ -16,6 +25,37 @@
...
+public_key_path = os.path.abspath(os.path.join(key_dir, 'public.pem'))
```
Comment (hannes-ucsc) [inline test/app_test_case.py]: Shouldn't be a global variable.

File: test/app_test_case.py [hunk 8]
```diff
@@ -16,6 +25,37 @@
...
+private_key_path = os.path.abspath(os.path.join(key_dir, 'private.pem'))
```
Comment (hannes-ucsc) [inline test/app_test_case.py]: Shouldn't be a global variable.

File: test/app_test_case.py [hunk 9]
```diff
@@ -16,6 +25,37 @@
...
+
+
+def generate_test_keys():
```
Comment (hannes-ucsc) [inline test/app_test_case.py]: This should be a method or property in the test case. The property should generate the keypair if necessary and return the path to it.

File: test/app_test_case.py [hunk 10]
```diff
@@ -16,6 +25,37 @@
...
+    if not os.path.exists(private_key_path):
+        call(['openssl', 'genrsa', '-out', private_key_path, '2048'])
+        call(['openssl', 'rsa', '-in', private_key_path, '-outform', 'PEM', '-pubout', '-out', public_key_path])
+
+
+def get_test_keys():
```
Comment (hannes-ucsc) [inline test/app_test_case.py]: Same.

File: test/app_test_case.py [hunk 11]
```diff
@@ -16,6 +25,37 @@
...
+        public_key = serialization.load_pem_public_key(
+            key_file.read(),
+            backend=default_backend()
+        )
+
+    return private_key, public_key
+
+
+def remove_test_keys():
+    os.unlink(public_key_path)
```
Comment (hannes-ucsc) [inline test/app_test_case.py]: Method in a test. I don't understand why this deviates from a well established pattern of test classes and methods to global functions.

File: lambdas/service/app.py [hunk 12]
```diff
@@ -808,13 +842,6 @@ def files_proxy(uuid):
         dss_response.raise_for_status()
 
 
-def file_url(uuid, **params):
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]: Inconsequential change. Just changes the order of two methods. Remove.

File: lambdas/service/app.py [hunk 13]
```diff
@@ -824,6 +851,58 @@ def self_url(endpoint_path=None):
     return retry_url
 
 
+def file_url(uuid, **params):
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]: Inconsequential change. Just changes the order of two methods. Remove.

File: src/azul/security/verifier.py (hunk 3 above)
Comment (shiroyuki) [reply src/azul/security/verifier.py]: This is moved to `azul.security.authenticator.Authenticator.session`.

File: src/azul/security/verifier.py (hunk 5 above)
Comment (shiroyuki) [reply src/azul/security/verifier.py]: Updated (inline)

File: test/app_test_case.py (hunk 9 above)
Comment (shiroyuki) [reply test/app_test_case.py]: Currently, this is part of a helper class, called `TestKeyManager` in `app_test_case.py`. I am thinking of either leaving it (to keep it isolated from growing `LocalAppTestCase`) as it is, or merging it into `LocalAppTestCase`. Which approach would you prefer, @hannes-ucsc ?

File: test/app_test_case.py (hunk 10 above)
Comment (shiroyuki) [reply test/app_test_case.py]: Currently, this is part of a helper class, called `TestKeyManager` in `app_test_case.py`.

File: test/app_test_case.py (hunk 11 above)
Comment (shiroyuki) [reply test/app_test_case.py]: Currently, this is part of a helper class, called `TestKeyManager` in `app_test_case.py`.

File: lambdas/service/app.py (hunk 13 above)
Comment (shiroyuki) [reply lambdas/service/app.py]: It was a rebasing issue. I moved it back.

File: lambdas/service/app.py (hunk 12 above)
Comment (shiroyuki) [reply lambdas/service/app.py]: It was a rebasing issue. It has been moved back to this location.

File: lambdas/indexer/.gitignore [hunk 14]
```diff
@@ -3,4 +3,5 @@
 /.chalice/config.json
 /.chalice/deployed/*.json
 /lambda-policy.json
-/vendor/azul_changes.py
+/vendor/*
```
Comment (hannes-ucsc) [inline lambdas/indexer/.gitignore]: Next time you push, please push a history with two commits. First commit should have the changes to these `.gitignore` files. Second commit is everything else.

File: environment [hunk 15]
```diff
@@ -145,6 +145,9 @@ azul_vars=(
     # string, only bundles whose UUID starts with the specified string will be
     # indexed.
     azul_dss_query_prefix
+
+    # Hostname of Fusillade (Authentication Broker Service)
```
Comment (hannes-ucsc) [inline environment]: This isn't actually just a host name. It's a URL prefix.

File: test/app_test_case.py [hunk 16]
```diff
@@ -65,6 +120,7 @@ def base_url(self):
     @classmethod
     def setUpClass(cls):
         super().setUpClass()
+        TestKeyManager.generate_test_keys()
```
Comment (hannes-ucsc) [inline test/app_test_case.py]: Why do we need to do this for every test case? Aren't the keys only needed for the authentication test cases?

File: test/app_test_case.py [hunk 17]
```diff
@@ -108,3 +162,54 @@ def tearDown(self):
         self.server_thread.join(timeout=10)
         if self.server_thread.is_alive():
             self.fail('Thread is still alive after joining')
+
+    def set_up_fake_responses(self):
+        # This requires to be invoked in a method within the context of
+        # ``responses.activate`` or ``responses.RequestsMock()``.
```
Comment (hannes-ucsc) [inline test/app_test_case.py]: Consider making this method a decorator that calls `responses.activate` so as to enforce the dependency.

Also take a look at `test/retorts.py`. It works around a problem that we may also run into again when once we add auth to the cart endpoints AND run tests against those.

Instead of creating another decorator, the best approach may be to extend `ResponsesHelper` in `retorts.py` and to override the `__enter__` method in order to set up the mocked auth endpoint responses. 

If the decorator approach works for now, I'd be happy with that, too.

Also, prefer the term `mock` over `fake`.

File: test/app_test_case.py (hunk 16 above)
Comment (shiroyuki) [reply test/app_test_case.py]: No, we don't. I'll change this into making the keys on demand.

Comment (shiroyuki) [reply test/app_test_case.py]: In the next set of commits, the test keys will be generated just right before either one of the keys are requested.

File: test/app_test_case.py (hunk 17 above)
Comment (shiroyuki) [reply test/app_test_case.py]: The setup for mock responses is moved to `AuthResponsesHelper` in `test/retorts.py` and the test case is split up into `LocalAppTestCase` and `AuthLocalAppTestCase` (still in the same file).

File: lambdas/indexer/.gitignore (hunk 14 above)
Comment (hannes-ucsc) [reply lambdas/indexer/.gitignore]: This hasn't been addressed.

File: test/retorts.py [hunk 18]
```diff
@@ -51,3 +65,114 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
         self.request_mock.passthru_prefixes = self.passthru_prefixes
+
+
+class AuthResponseHelper(ResponsesHelper):
+    def __init__(self, passthru_url: str = None, request_mock: responses.RequestsMock = None) -> None:
+        super().__init__(request_mock)
+        self.passthru_url = passthru_url
+
+    def __enter__(self):
+        context = super().__enter__()
```
Comment (hannes-ucsc) [inline test/retorts.py]: You should return `context`.

Comment (shiroyuki) [reply test/retorts.py]: @hannes-ucsc `return context` is already on [line 103](https://github.com/DataBiosphere/azul/pull/685/files/5378ac6a7c9d1ca0b61f711b7d34e0483fdaf9e7#diff-225b710c3efb09eae6db97fdd83cf528R103). If it isn't what it should be, could you clarify what you meant by returning `context`?

File: lambdas/indexer/.gitignore (hunk 14 above)
Comment (shiroyuki) [reply lambdas/indexer/.gitignore]: @hannes-ucsc This change was made when we try to streamline `make deploy` to install additional wheels (b7690914). Since then, we identified that the compatibility between two Python versions (3.6 and 3.7) caused the issue and we addressed that by ensuring that everyone must use Python 3.6 for the deployment.

I think this change seems to be unnecessary at this point and I think of reverting this change. What do you think?

File: test/app_test_case.py [hunk 19]
```diff
@@ -78,10 +78,7 @@ def tearDownClass(cls):
 
     def setUp(self):
         super().setUp()
-        log.debug("Setting up tests")
```
Comment (hannes-ucsc) [inline test/app_test_case.py]: Change makes sense, but should have been done in a separate commit.

File: test/retorts.py (hunk 18 above)
Comment (hannes-ucsc) [reply test/retorts.py]: You are right. My mistake.
